wire-ios/Wire-iOS/Sources/UserInterface/Conversation/Content/MessagePresenter.swift

```diff
@@ -204,28 +207,46 @@ final class MessagePresenter: NSObject {
 
     func viewController(forImageMessagePreview message: ZMConversationMessage, actionResponder delegate: MessageActionResponder) -> UIViewController? {
         guard Message.isImage(message),
-            message.imageMessageData != nil else {
-                return nil
+              message.imageMessageData != nil else {
+            return nil
         }
 
         return imagesViewController(for: message, actionResponder: delegate, isPreviewing: true)
     }
 
     // MARK: - Pass
 
-    func createAddPassesViewController(fileMessageData: ZMFileMessageData) -> PKAddPassesViewController? {
-        guard let fileURL = fileMessageData.fileURL,
-            let passData = try? Data.init(contentsOf: fileURL) else {
-                return nil
+    @MainActor
+    func openPassesViewController(fileMessageData: ZMFileMessageData) async {
+        guard PKAddPassesViewController.canAddPasses() else { return }
+
+        guard let viewController = try? await makePassesViewController(fileMessageData: fileMessageData) else {
+            return // ignore error
         }
 
-        guard let pass = try? PKPass.init(data: passData) else { return nil }
+        targetViewController?.present(viewController, animated: true)
+    }
 
-        if PKAddPassesViewController.canAddPasses() {
-            return PKAddPassesViewController(pass: pass)
-        } else {
-            return nil
+    @MainActor
+    func makePassesViewController(fileMessageData: ZMFileMessageData) async throws -> PKAddPassesViewController? {
+        guard let fileURL = fileMessageData.fileURL else {
+            throw MessagePresenterError.missingFileURL
         }
+
+        let pass = try await loadPass(fileURL: fileURL)
+        return PKAddPassesViewController(pass: pass)
+    }
+
+    private func loadPass(fileURL: URL) async throws -> PKPass {
+        // Without MainActor attribute, this runs on a background queue.
+        let task = Task.detached(
+            priority: .userInitiated,
+            operation: {
+                let passData = try Data(contentsOf: fileURL)
+                return try PKPass(data: passData)
+            }
+        )
+        return try await task.value
```

Comment:
```suggestion
@MainActor
private func loadPass(fileURL: URL) async throws -> PKPass {
    let data = try await Task {
        try Data(contentsOf: fileURL)
    }.value
    return try PKPass(data: data)
}
```

I'm not sure if this runs the IO-part on a background queue and the `PKPass.init` on the main queue, but you could give it a try.

Reply:
I would try it, but I don't want to change my branch now.

Reply:
Why do you want to change creating `PKPass` on a background queue to main queue? In our tests it worked this way and it's better to have less work on main

Reply:
`PKPass.init` must be run on the main queue, see the Xcode warning in the PR description.